File: tests/benchmarks/fixtures/sample.py
"""Sample Python module for benchmarking scissorhands AST operations."""

import os
import sys
import json
import logging
from typing import Optional, List, Dict, Any, Tuple
from dataclasses import dataclass, field
from pathlib import Path
from datetime import datetime, timedelta

# Configuration constants
DEFAULT_TIMEOUT = 5000
MAX_RETRIES = 3
BATCH_SIZE = 100

logger = logging.getLogger(__name__)


@dataclass
class Config:
    host: str = "localhost"
    port: int = 8080
    debug: bool = False
    max_connections: int = 10


@dataclass
class User:
    id: str
    name: str
    email: str
    role: str = "user"
    created_at: datetime = field(default_factory=datetime.now)


@dataclass
class QueryResult:
    data: List[Any] = field(default_factory=list)
    total: int = 0
    page: int = 1
    page_size: int = 20


class DatabaseConnection:
    """Manages database connections with retry logic."""

    def __init__(self, config: Config):
        self.config = config
        self.connected = False
        self.retry_count = 0
        print(f"DatabaseConnection created with config: {config}")

    def connect(self) -> None:
        print(f"Connecting to {self.config.host}:{self.config.port}...")
        try:
            self.connected = True
            self.retry_count = 0
            print("Connected successfully")
        except Exception as e:
            self.retry_count += 1
            print(f"Connection failed, retry {self.retry_count}/{MAX_RETRIES}")
            if self.retry_count < MAX_RETRIES:
                self.connect()
            raise e

    def disconnect(self) -> None:
        print("Disconnecting...")
        self.connected = False
        print("Disconnected")

    def is_connected(self) -> bool:
        return self.connected


class UserRepository:
    """Repository pattern for user data access."""

    def __init__(self, db: DatabaseConnection):
        self.db = db
        self.cache: Dict[str, User] = {}
        print("UserRepository initialized")

    def find_by_id(self, user_id: str) -> Optional[User]:
        print(f"Finding user by id: {user_id}")
        cached = self.cache.get(user_id)
        if cached:
            print(f"Cache hit for user {user_id}")
            return cached
        print(f"Cache miss for user {user_id}, querying database")
        return None

    def find_by_email(self, email: str) -> Optional[User]:
        print(f"Finding user by email: {email}")
        for user in self.cache.values():
            if user.email == email:
                return user
        return None

    def find_all(self, page: int = 1, page_size: int = 20) -> QueryResult:
        print(f"Fetching users page {page}, size {page_size}")
        users = list(self.cache.values())
        start = (page - 1) * page_size
        data = users[start : start + page_size]
        return QueryResult(data=data, total=len(users), page=page, page_size=page_size)

    def create(self, user: User) -> User:
        print(f"Creating user: {user.name}")
        self.cache[user.id] = user
        print(f"User {user.name} created successfully")
        return user

    def update(self, user_id: str, **updates: Any) -> Optional[User]:
        print(f"Updating user {user_id}")
        existing = self.cache.get(user_id)
        if not existing:
            print(f"User {user_id} not found for update")
            return None
        for key, value in updates.items():
            setattr(existing, key, value)
        print(f"User {user_id} updated successfully")
        return existing

    def delete(self, user_id: str) -> bool:
        print(f"Deleting user {user_id}")
        if user_id in self.cache:
            del self.cache[user_id]
            print(f"User {user_id} deleted")
            return True
        print(f"User {user_id} not found")
        return False

    def get_cache_size(self) -> int:
        return len(self.cache)


class EventBus:
    """Simple event bus for publishing and subscribing to events."""

    def __init__(self):
        self.handlers: Dict[str, List[Any]] = {}
        self.event_log: List[Dict[str, Any]] = []
        print("EventBus initialized")

    def register(self, event: str, handler: Any) -> None:
        print(f"Registering handler for event: {event}")
        if event not in self.handlers:
            self.handlers[event] = []
        self.handlers[event].append(handler)

    def dispatch(self, event: str, data: Any = None) -> None:
        print(f"Dispatching event: {event}")
        self.event_log.append({"event": event, "timestamp": datetime.now()})
        handlers = self.handlers.get(event, [])
        for handler in handlers:
            try:
                handler(event, data)
            except Exception as e:
                print(f"Handler error for event {event}: {e}")

    def get_event_log(self) -> List[Dict[str, Any]]:
        return list(self.event_log)


def validate_email(email: str) -> bool:
    print(f"Validating email: {email}")
    import re
    pattern = r"^[^\s@]+@[^\s@]+\.[^\s@]+$"
    return bool(re.match(pattern, email))


def generate_id() -> str:
    import uuid
    return str(uuid.uuid4())[:15]


def format_user(user: User) -> str:
    print(f"Formatting user: {user.name}")
    return f"{user.name} <{user.email}> ({user.role})"


def load_config(path: str) -> Config:
    print(f"Loading config from {path}")
    with open(path) as f:
        data = json.load(f)
    config = Config(**data)
    print(f"Config loaded: {config}")
    return config


def save_config(path: str, config: Config) -> None:
    print(f"Saving config to {path}")
    data = {
        "host": config.host,
        "port": config.port,
        "debug": config.debug,
        "max_connections": config.max_connections,
    }
    with open(path, "w") as f:
        json.dump(data, f, indent=2)
    print("Config saved")


def process_items(items: List[Any], batch_size: int = BATCH_SIZE) -> List[List[Any]]:
    print(f"Processing {len(items)} items in batches of {batch_size}")
    batches = []
    for i in range(0, len(items), batch_size):
        batches.append(items[i : i + batch_size])
    print(f"Created {len(batches)} batches")
    return batches


def retry_operation(operation: Any, max_retries: int = MAX_RETRIES, delay: float = 1.0) -> Any:
    print(f"Starting operation with {max_retries} retries")
    for attempt in range(1, max_retries + 1):
        try:
            result = operation()
            print(f"Operation succeeded on attempt {attempt}")
            return result
        except Exception as e:
            print(f"Attempt {attempt} failed: {e}")
            if attempt == max_retries:
                raise
            import time
            time.sleep(delay * attempt)
    raise RuntimeError("Unreachable")

File: tests/benchmarks/fixtures/sample.ts
import { readFile, writeFile } from 'node:fs/promises';
import { join, resolve, dirname } from 'node:path';
import { EventEmitter } from 'node:events';

// Configuration constants
const DEFAULT_TIMEOUT = 5000;
const MAX_RETRIES = 3;
const BATCH_SIZE = 100;

interface Config {
  host: string;
  port: number;
  debug: boolean;
  maxConnections: number;
}

interface User {
  id: string;
  name: string;
  email: string;
  role: 'admin' | 'user' | 'guest';
  createdAt: Date;
}

interface QueryResult<T> {
  data: T[];
  total: number;
  page: number;
  pageSize: number;
}

type EventHandler = (event: string, data: unknown) => void;

class DatabaseConnection {
  private config: Config;
  private connected: boolean = false;
  private retryCount: number = 0;

  constructor(config: Config) {
    this.config = config;
    console.log('DatabaseConnection created with config:', config);
  }

  async connect(): Promise<void> {
    console.log(`Connecting to ${this.config.host}:${this.config.port}...`);
    try {
      this.connected = true;
      this.retryCount = 0;
      console.log('Connected successfully');
    } catch (error) {
      this.retryCount++;
      console.log(`Connection failed, retry ${this.retryCount}/${MAX_RETRIES}`);
      if (this.retryCount < MAX_RETRIES) {
        await this.connect();
      }
      throw error;
    }
  }

  async disconnect(): Promise<void> {
    console.log('Disconnecting...');
    this.connected = false;
    console.log('Disconnected');
  }

  isConnected(): boolean {
    return this.connected;
  }
}

class UserRepository {
  private db: DatabaseConnection;
  private cache: Map<string, User> = new Map();

  constructor(db: DatabaseConnection) {
    this.db = db;
    console.log('UserRepository initialized');
  }

  async findById(id: string): Promise<User | undefined> {
    console.log(`Finding user by id: ${id}`);
    const cached = this.cache.get(id);
    if (cached) {
      console.log(`Cache hit for user ${id}`);
      return cached;
    }
    console.log(`Cache miss for user ${id}, querying database`);
    return undefined;
  }

  async findByEmail(email: string): Promise<User | undefined> {
    console.log(`Finding user by email: ${email}`);
    for (const user of this.cache.values()) {
      if (user.email === email) {
        return user;
      }
    }
    return undefined;
  }

  async findAll(page: number = 1, pageSize: number = 20): Promise<QueryResult<User>> {
    console.log(`Fetching users page ${page}, size ${pageSize}`);
    const users = Array.from(this.cache.values());
    const start = (page - 1) * pageSize;
    const data = users.slice(start, start + pageSize);
    return {
      data,
      total: users.length,
      page,
      pageSize,
    };
  }

  async create(user: User): Promise<User> {
    console.log(`Creating user: ${user.name}`);
    this.cache.set(user.id, user);
    console.log(`User ${user.name} created successfully`);
    return user;
  }

  async update(id: string, updates: Partial<User>): Promise<User | undefined> {
    console.log(`Updating user ${id}`);
    const existing = this.cache.get(id);
    if (!existing) {
      console.log(`User ${id} not found for update`);
      return undefined;
    }
    const updated = { ...existing, ...updates };
    this.cache.set(id, updated);
    console.log(`User ${id} updated successfully`);
    return updated;
  }

  async delete(id: string): Promise<boolean> {
    console.log(`Deleting user ${id}`);
    const deleted = this.cache.delete(id);
    console.log(`User ${id} ${deleted ? 'deleted' : 'not found'}`);
    return deleted;
  }

  getCacheSize(): number {
    return this.cache.size;
  }
}

class EventBus extends EventEmitter {
  private handlers: Map<string, EventHandler[]> = new Map();
  private eventLog: Array<{ event: string; timestamp: Date }> = [];

  constructor() {
    super();
    console.log('EventBus initialized');
  }

  register(event: string, handler: EventHandler): void {
    console.log(`Registering handler for event: ${event}`);
    const existing = this.handlers.get(event) ?? [];
    existing.push(handler);
    this.handlers.set(event, existing);
  }

  dispatch(event: string, data: unknown): void {
    console.log(`Dispatching event: ${event}`);
    this.eventLog.push({ event, timestamp: new Date() });
    const handlers = this.handlers.get(event) ?? [];
    for (const handler of handlers) {
      try {
        handler(event, data);
      } catch (error) {
        console.log(`Handler error for event ${event}:`, error);
      }
    }
  }

  getEventLog(): Array<{ event: string; timestamp: Date }> {
    return [...this.eventLog];
  }
}

function validateEmail(email: string): boolean {
  console.log(`Validating email: ${email}`);
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  return emailRegex.test(email);
}

function generateId(): string {
  return Math.random().toString(36).substring(2, 15);
}

function formatUser(user: User): string {
  console.log(`Formatting user: ${user.name}`);
  return `${user.name} <${user.email}> (${user.role})`;
}

async function loadConfig(path: string): Promise<Config> {
  console.log(`Loading config from ${path}`);
  const content = await readFile(path, 'utf-8');
  const config = JSON.parse(content) as Config;
  console.log('Config loaded:', config);
  return config;
}

async function saveConfig(path: string, config: Config): Promise<void> {
  console.log(`Saving config to ${path}`);
  const content = JSON.stringify(config, null, 2);
  await writeFile(path, content, 'utf-8');
  console.log('Config saved');
}

function processItems<T>(items: T[], batchSize: number = BATCH_SIZE): T[][] {
  console.log(`Processing ${items.length} items in batches of ${batchSize}`);
  const batches: T[][] = [];
  for (let i = 0; i < items.length; i += batchSize) {
    batches.push(items.slice(i, i + batchSize));
  }
  console.log(`Created ${batches.length} batches`);
  return batches;
}

async function retryOperation<T>(
  operation: () => Promise<T>,
  maxRetries: number = MAX_RETRIES,
  delay: number = 1000,
): Promise<T> {
  console.log(`Starting operation with ${maxRetries} retries`);
  for (let attempt = 1; attempt <= maxRetries; attempt++) {
    try {
      const result = await operation();
      console.log(`Operation succeeded on attempt ${attempt}`);
      return result;
    } catch (error) {
      console.log(`Attempt ${attempt} failed:`, error);
      if (attempt === maxRetries) throw error;
      await new Promise((r) => setTimeout(r, delay * attempt));
    }
  }
  throw new Error('Unreachable');
}

export {
  DatabaseConnection,
  UserRepository,
  EventBus,
  validateEmail,
  generateId,
  formatUser,
  loadConfig,
  saveConfig,
  processItems,
  retryOperation,
  DEFAULT_TIMEOUT,
  MAX_RETRIES,
  BATCH_SIZE,
};

export type { Config, User, QueryResult, EventHandler };
